src/pkg/zoci/pull.go

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package zoci contains functions for interacting with Zarf packages stored in OCI registries.
+package zoci
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"slices"
+
+	"github.com/defenseunicorns/zarf/src/pkg/layout"
+	"github.com/defenseunicorns/zarf/src/pkg/oci"
+	"github.com/defenseunicorns/zarf/src/pkg/transform"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
+	"github.com/defenseunicorns/zarf/src/types"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content/file"
+)
+
+var (
+	// PackageAlwaysPull is a list of paths that will always be pulled from the remote repository.
+	PackageAlwaysPull = []string{layout.ZarfYAML, layout.Checksums, layout.Signature}
+)
+
+// PullPackage pulls the package from the remote repository and saves it to the given path.
+//
+// layersToPull is an optional parameter that allows the caller to specify which layers to pull.
+//
+// The following layers will ALWAYS be pulled if they exist:
+//   - zarf.yaml
+//   - checksums.txt
+//   - zarf.yaml.sig
+func (r *Remote) PullPackage(ctx context.Context, destinationDir string, concurrency int, layersToPull ...ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	isPartialPull := len(layersToPull) > 0
+	r.Log().Debug(fmt.Sprintf("Pulling %s", r.Repo().Reference))
+
+	manifest, err := r.FetchRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if isPartialPull {
+		for _, path := range PackageAlwaysPull {
+			desc := manifest.Locate(path)
+			layersToPull = append(layersToPull, desc)
+		}
+	} else {
+		layersToPull = append(layersToPull, manifest.Layers...)
+	}
+	layersToPull = append(layersToPull, manifest.Config)
+
+	// Create a thread to update a progress bar as we save the package to disk
+	doneSaving := make(chan error)
+	successText := fmt.Sprintf("Pulling %q", helpers.OCIURLPrefix+r.Repo().Reference.String())
+
+	layerSize := oci.SumDescsSize(layersToPull)
+	go utils.RenderProgressBarForLocalDirWrite(destinationDir, layerSize, doneSaving, "Pulling", successText)
+
+	dst, err := file.New(destinationDir)
+	if err != nil {
+		return nil, err
+	}
+	defer dst.Close()
+
+	copyOpts := r.GetDefaultCopyOpts()
+	copyOpts.Concurrency = concurrency
+
+	err = r.CopyToTarget(ctx, layersToPull, dst, copyOpts)
+	doneSaving <- err
+	<-doneSaving
+	return layersToPull, err
+}
+
+// LayersFromRequestedComponents returns the descriptors for the given components from the root manifest.
+// It also retrieves the descriptors for all image layers that are required by the components.
+//
+// It also respects the `required` flag on components, and will retrieve all necessary layers for required components.
+func (r *Remote) LayersFromRequestedComponents(ctx context.Context, requestedComponents []string) (layers []ocispec.Descriptor, err error) {
```

Comment:
depending upon which one (`filters` or this PR) gets merged first, will need to keep an eye on this func to ensure regression does not happen

Reply:
Gotcha, plan is to get this PR in first